e2e/testcases/namespace_selectors_test.go

```diff
@@ -59,73 +99,226 @@ func TestNamespaceSelectorHierarchicalFormat(t *testing.T) {
+			testpredicates.ResourceGroupHasNoObjects(unselectedResourcesWithShoestoreNSSOnly),
+		}); err != nil {
+		nt.T.Fatal(err)
+	}
 	if err := nt.WatchForAllSyncs(); err != nil {
 		nt.T.Fatal(err)
 	}
-	if err := nt.Validate(bookstoreCM.Name, bookstoreNS, &corev1.ConfigMap{}); err != nil {
+	validateSelectedAndUnselectedResources(nt,
+		append(selectedResourcesWithShoestoreNSSOnly),
+		append(unselectedResourcesWithShoestoreNSSOnly),
+	)
+
+	nt.Logger.Info("Update Namespace's label to make it selected again, resources should be selected")
+	nt.MustMergePatch(bookstoreNamespace, fmt.Sprintf(`{"metadata":{"labels":{"app":"%s"}}}`, bookstoreNS))
+
+	if err := nt.Watcher.WatchObject(kinds.ResourceGroup(),
+		configsync.RootSyncName,
+		configsync.ControllerNamespace,
+		[]testpredicates.Predicate{
+			testpredicates.ResourceGroupHasObjects(selectedResourcesWithBookstoreNSSAndShoestoreNSS),
+			testpredicates.ResourceGroupHasNoObjects(unselectedResourcesWithBookstoreNSSAndShoestoreNSS),
+		}); err != nil {
 		nt.T.Fatal(err)
 	}
-	if err := nt.Validate(bookstoreRQ.Name, bookstoreNS, &corev1.ResourceQuota{}, resourceQuotaHasHardPods(nt, "1")); err != nil {
+	if err := nt.WatchForAllSyncs(); err != nil {
 		nt.T.Fatal(err)
 	}
-	if err := nt.Validate(shoestoreCM.Name, shoestoreNS, &corev1.ConfigMap{}); err != nil {
+	validateSelectedAndUnselectedResources(nt,
+		selectedResourcesWithBookstoreNSSAndShoestoreNSS,
+		unselectedResourcesWithBookstoreNSSAndShoestoreNSS,
+	)
+
+	nt.Logger.Info("Stop the admission webhook so the Namespace can be deleted")
+	// If the webhook is running, the namespace DELETE call will fail because the
+	// underlying resources cannot be deleted due to the admission-webhook.
+	nomostest.StopWebhook(nt)
+	nt.Logger.Info("Delete Namespace, resources should NOT be selected")
+	if err := nt.KubeClient.Delete(bookstoreNamespace); err != nil {
+		t.Fatal(err)
+	}
+
+	nt.Logger.Info("Watching the ResourceGroup object until unselected resources are removed from the inventory")
+	if err := nt.Watcher.WatchObject(kinds.ResourceGroup(),
+		configsync.RootSyncName,
+		configsync.ControllerNamespace,
+		[]testpredicates.Predicate{
+			testpredicates.ResourceGroupHasObjects(selectedResourcesWithShoestoreNSSOnly),
+			testpredicates.ResourceGroupHasNoObjects(unselectedResourcesWithShoestoreNSSOnly),
+		}); err != nil {
 		nt.T.Fatal(err)
 	}
-	if err := nt.ValidateNotFound(bookstoreCM.Name, shoestoreNS, &corev1.ConfigMap{}); err != nil {
+	if err := nt.WatchForAllSyncs(); err != nil {
 		nt.T.Fatal(err)
 	}
-	if err := nt.ValidateNotFound(bookstoreRQ.Name, shoestoreNS, &corev1.ResourceQuota{}); err != nil {
+	validateSelectedAndUnselectedResources(nt,
+		append(selectedResourcesWithShoestoreNSSOnly),
+		append(unselectedResourcesWithShoestoreNSSOnly),
+	)
+
+	nt.Logger.Info("Update NamespaceSelector to use static mode")
+	bookstoreNSS.Spec.Mode = v1.NSSelectorStaticMode
```

Comment:
What happens if the mode is switched from `dynamic` -> `static` while there are currently resources in a "dynamic" namespace?

That seems like an interesting scenario, but I don't think the current test case covers that because the dynamic namespace is deleted before switching from `dynamic` -> `static`.

Reply:
Moves the case right after the resources are dynamically selected, and then change it back to `dynamic` mode to cover more scenarios.